akka-http-core/src/main/scala/akka/http/scaladsl/model/headers/headers.scala

```diff
@@ -57,6 +61,60 @@ abstract class CustomHeader extends jm.headers.CustomHeader {
   final def render[R <: Rendering](r: R): r.type = r ~~ name ~~ ':' ~~ ' ' ~~ value
 }
 
+abstract class ModeledCustomHeaderCompanion[H <: ModeledCustomHeader[H]] {
+  def name: String
+  def lowercaseName: String = name.toRootLowerCase
+
+  def parse(value: String): Option[H]
+
+  def apply(value: String): H =
+    parse(value).getOrElse(throw new RuntimeException("Unable to construct custom header by parsing value:" + value))
+
+  def unapply(h: HttpHeader): Option[String] = h match {
+    case _: RawHeader    ⇒ if (h.lowercaseName == lowercaseName) Some(h.value) else None
+    case _: CustomHeader ⇒ if (h.lowercaseName == lowercaseName) Some(h.value) else None
+    case _               ⇒ None
+  }
+
+}
+
+/**
+ * Superclass for user-defined custom headers defined by implementing `name` and `value`.
+ */
+abstract class ModeledCustomHeader[H <: ModeledCustomHeader[H]] extends CustomHeader { this: H ⇒
+  def companion: ModeledCustomHeaderCompanion[H]
+
+  final override def name = companion.name
+  final override def lowercaseName: String = name.toRootLowerCase
+}
+
+///**
+// * Superclass for user-defined custom headers defined by implementing `name` and `value`.
+// */
+//abstract class CustomHeader[Self <: CustomHeader[Self]] extends jm.headers.CustomHeader {
+//  def companion: CustomHeaderCompanion[Self]
+//
+//  /** Override to return true if this header shouldn't be rendered */
+//  def suppressRendering: Boolean = false
+//
+//  def name: String = companion.name
+//  def lowercaseName: String = name.toRootLowerCase
+//  final def render[R <: Rendering](r: R): r.type = r ~~ name ~~ ':' ~~ ' ' ~~ value
+//}
+abstract class CustomHeaderCompanion[H <: CustomHeader] {
+  def name: String
+  def lowercaseName: String = name.toRootLowerCase
+
+  def parse(value: String): Option[H]
+  def apply(value: String): H = parse(value).getOrElse(throw new Exception("Whoops"))
```

Comment:
Maybe `RuntimeException` rather than `Exception`?
And an error message that shows the offending input?